third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc

```diff
@@ -2340,11 +2363,38 @@ static void RunScript(v8::Isolate* isolate, v8::Local<v8::Context> context, cons
   }
 }
 
+static void fromJsRegisterRootFrameScript(
+    const v8::FunctionCallbackInfo<v8::Value>& args) {
+  CHECK(args.Length() == 1 && args[0]->IsString() &&
+        "must be called with a single string");
+
+  v8::Isolate* isolate = args.GetIsolate();
+
+  auto contextGroupId = GetCurrentContextGroupIdForIsolate(isolate);
+
+  if (!contextGroupId.has_value()) {
+    recordreplay::Warning(
+        "fromJsRegisterRootFrameScript - no valid context id");
+    args.GetReturnValue().Set(v8::Boolean::New(isolate, false));
+    return;
+  }
+
+  auto context = isolate->GetCurrentContext();
+
+  v8::String::Utf8Value script(args.GetIsolate(), args[0]);
+  RunScript(isolate, context, *script, kInternalScriptURL);
```

Comment:
I like the idea of running the registered script right away - this allows it to be registered at any point in time. It's not quite needed right now but this is a general-purpose API.

However, at the moment, it also means that for the *first* root frame this stuff gets executed twice because apparently immediately called `Target.evaluatePrivileged` manages to get executed before `OnRootFrameInitAfterCheckpoint`. I don't exactly know where the executed commands are flushed for the first time so I'm not sure how to remedy this. This also doesn't quite cause any issues with the logic we are currently using this for - but it would be nice to fix this.

Reply:
Hmm, when we run the instrument engine preloads it will be at the first checkpoint, so wouldn't it be expected that OnRootFrameInitAfterCheckpoint runs after that point?  It seems like this RunScript call could be removed, we could assert that the root scripts are only added at the first checkpoint or just have the semantics be that the script will only be called on future root frames.

Reply:
> Hmm, when we run the instrument engine preloads it will be at the first checkpoint, so wouldn't it be expected that OnRootFrameInitAfterCheckpoint runs after that point?

Yeah, that's probably true and that's what I expected based on the observed behavior. It's just that I couldn't quite tell from the code alone where those preloads would be executed. What makes them executed at the first checkpoint? Is it `InitializeRecordReplayAfterCheckpoint()`?

> or just have the semantics be that the script will only be called on future root frames.

Yeah, I'll probably just roll with this.